Complete first sprint. Step second

uncledeni opened the pull request:
Add Redux and React-DnD to project

Rework data store from useState to Redux store; connect 'store' Provider to root index.js file; activate REDUX_DEVTOOLS_EXTENSION_COMPOSE__;

Create following several actions and reducers files; Create 'rootReducer' as combineReducers; connect 'rootReducer' to root index.js file;

Add 'DndProvider' to 'MainMenu' page;
make DnD ingredients add from list to constructor; make DnD elements sort in constructor;

Add 'postOrder' api;
create few utils 'checks';

uncledeni committed:
Add Redux and React-DnD to project

Rework data store from useState to Redux store; connect 'store' Provider to root index.js file;
activate REDUX_DEVTOOLS_EXTENSION_COMPOSE__;

Create following several actions and reducers files;
Create 'rootReducer' as combineReducers; connect 'rootReducer' to root index.js file;

Add 'DndProvider' to 'MainMenu' page;
make DnD ingredients add from list to constructor;
make DnD elements sort in constructor;

Add 'postOrder' api;
create few utils 'checks';

gennady-bars commented on src/shared/services/actions/order-details.js:
```diff
@@ -0,0 +1,27 @@
+                    type: GET_ORDER_DETAILS_FAILED
+                });
+            }
+        });
```
Для любых запросов на сервер в **самом конце цепочки** запроса нужно осуществлять проверку на возможные ошибки с помощью `catch`.

**Нужно исправить это везде по коду в конце запросов.**

gennady-bars commented on src/shared/services/actions/burger-ingredients.js:
```diff
@@ -0,0 +1,29 @@
+                    type: GET_BURGER_INGREDIENTS_FAILED
+                });
+            }
+        });
```
Для любых запросов на сервер в **самом конце цепочки** запроса нужно осуществлять проверку на возможные ошибки с помощью `catch`.

gennady-bars commented on src/shared/services/actions/burger-ingredients.js:
```diff
@@ -0,0 +1,29 @@
+            type: GET_BURGER_INGREDIENTS_REQUEST
+        });
+        getDataService().then(res => {
+            if (res && res.success) {
```
**Можно лучше**


Если интересно, то можно создать **универсальную** функцию запроса `request` , которая будет содержать в себе уже **базовый урл**, проверку на **ok** и проверку на `success`, чтобы не дублировать их в каждом запросе.

```js

// 1 раз объявляем базовый урл
export const BASE_URL = "https://norma.nomoreparties.space/api/";

// создаем функцию проверки ответа на `ok`
const checkResponse = (res) => {
  if (res.ok) {
    return res.json();
  }
  // не забываем выкидывать ошибку, чтобы она попала в `catch`
  return Promise.reject(`Ошибка ${res.status}`);
};

// создаем функцию проверки на `success`
const checkSuccess = (res) => {
  if (res && res.success) {
    return res;
  }
  // не забываем выкидывать ошибку, чтобы она попала в `catch`
  return Promise.reject(`Ответ не success: ${res}`);
};

// создаем универсальную фукнцию запроса с проверкой ответа и `success`
// В вызов приходит `endpoint`(часть урла, которая идет после базового) и опции
const request = (endpoint, options) => {
  // а также в ней базовый урл сразу прописывается, чтобы не дублировать в каждом запросе
  return fetch(`${BASE_URL}${endpoint}`, options)
    .then(checkResponse)
    .then(checkSuccess);
};

```


И теперь не нужно будет дублировать все это в каждом запросе. Нужно просто заменить в запросах все `fetch` на `request`, удалив лишние проверки и базовый урл. Все остальное останется без изменений

Код станет намного чище. Будет мелкий аналог `axios`

Пример запроса ингредиентов

```js
const getIngredients =  () => request("ingredients");

```

В `get`-запросах даже не нужно указывать 2й параметр. Он получится `undefined`, что без проблем для `fetch` (который внутри).

gennady-bars commented on src/shared/api/get-data-service.jsx:
```diff
@@ -8,4 +8,26 @@ export const getDataService = async () => {
     return response;
+}
+
+const ORDER_DATA = "https://norma.nomoreparties.space/api/orders";
```
`'https://norma.nomoreparties.space/api` везде повторяется в запросах. Нужно вынести в константу `BASE_URL` и импортировать везде в коде проекта, добавляя в конце изменяющуюся часть (эндпоит). Это обычная практика, чтобы одним движением можно было **сменить адрес**, но не менять эндпоинты.

gennady-bars commented on src/shared/api/get-data-service.jsx:
```diff
@@ -8,4 +8,26 @@ export const getDataService = async () => {
+                "ingredients": data
+            })
+        }
+    ).then(res => {
+        if (res.ok) {
+            return res.json();
+        }
+        return Promise.reject(`Ошибка HTTP: ${res.status}`);
```
```js
.then(res => {
                if (res.ok) {
                    return res.json();
                }
                return Promise.reject(`Ошибка ${res.status}`);
            })

```

У Вас повторяется код **проверки ответа** во всех запросах. Чтобы этого избежать, нужно сделать отдельную функцию `checkResponse`, которая будет проверять ответ от сервера:

```js
function checkResponse(res) {
        // тут проверка ответа
    }

```

Ее нужно поместить в папку `utils` и импортировать туда, где нужна проверка на `ok`.

и теперь можно удалить кучу ненужного повторного кода и заменить его этим:

```js
 .then(checkResponse)

```

Обратите внимание, что передается только **ссылка** на функцию. Не нужно ее вызывать. Она сама вызовется, так как в `then` нужно передавать именно **функцию**, а не **вызов** функции.

-----------------------------------------

**Можно лучше**

Можно сделать **универсальную** функцию запроса с **проверкой ответа**, чтобы не дублировать эту проверку в каждом запросе:

```js

  function request(url, options) {
    // принимает два аргумента: урл и объект опций, как и `fetch`
    return fetch(url, options).then(checkResponse)
  }

```

И теперь просто нужно заменить все `fetch` на `request` и убрать дублирование проверки на `ok`. Все остальное будет без изменений. Код станет чище

И даже можно поместить внутрь `baseUrl`, чтобы не дублировать его в каждом запросе. Тогда нужно будет передавать просто `endpoint` в вызов

gennady-bars commented on src/shared/api/get-data-service.jsx:
```diff
@@ -8,4 +8,26 @@ export const getDataService = async () => {
         return Promise.reject(`Ошибка HTTP: ${res.status}`);
```
У Вас повторяется код **проверки ответа** во всех запросах. Чтобы этого избежать, нужно сделать отдельную функцию `checkResponse`, которая будет проверять ответ от сервера:

gennady-bars commented on src/widgets/burger-constructor/burger-constructor.jsx:
```diff
@@ -54,40 +137,92 @@ const StackListElement = (props) => {
+    const renderIngredient = useCallback((ingredient, index) => {
+        return (
+            <StackListElement
+                key={ingredient.id}
```
`                key={ingredient.id}
`

Отлично, что указываете **уникальный** `id` для ингредиентов в заказе. Это гарантирует 100% логику `DnD`

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на **гитхабе** все комментарии к коду (нужно прокрутить вниз страницу там)

Работа проделана огромная:



* Отлично, что указываете **уникальный** `id` для ингредиентов в заказе. Это гарантирует 100% логику `DnD`
* Хорошая логика редакса и `DnD`

но есть некоторые недочеты:



* Для любых запросов на сервер в **самом конце цепочки** запроса нужно осуществлять проверку на возможные ошибки с помощью `catch`.
* `'https://norma.nomoreparties.space/api` везде повторяется в запросах. Нужно вынести в константу `BASE_URL` и импортировать везде в коде проекта, добавляя в конце изменяющуюся часть (эндпоит). Это обычная практика, чтобы одним движением можно было **сменить адрес**, но не менять эндпоинты. 
* У Вас повторяется код **проверки ответа** во всех запросах. Чтобы этого избежать, нужно сделать отдельную функцию `checkResponse`, которая будет проверять ответ от сервера:




**Можно лучше**

* Желательно **очищать конструктор** после успешного получения номера заказа с сервера в блоке `then`, чтобы пользователь мог следующий заказ сделать, не удаляя старые ингредиенты
* лучше не давать нажать на кнопку `Оформить заказ`, пока пользователь не добавил булку в заказ.
* лучше отображать **лоадер** (загрузку) при ожидании ответа от сервера с номером заказа

uncledeni committed:
Pull request fixes

make universal check functions for requests;
disable order button in case off empty 'bun' field;
add 'try/catch' block in actions;

** (try/catch блоки были добавлены, но алерты почему-то всё равно не выводятся. пробовал ставить их и на более низкоуровневые шаги - в блоки функций в файле catch, но и там эффекта алерта в catch не было. Подскажите, как в данном случае следует оформить эту обёртку)

gennady-bars commented on src/shared/services/actions/burger-ingredients.js:
```diff
@@ -13,17 +13,21 @@ export function getBurgerIngredients() {
-        });
+        try {
+            getDataService().then(res => {
+                if (res && res.success) {
```
` if (res && res.success) {` уже не нужно проверять, так как внутри `request` все делается уже

gennady-bars commented on src/shared/services/actions/order-details.js:
```diff
@@ -11,17 +11,21 @@ export function getOrderDetails(data) {
-        });
+        try {
+            postOrder(data).then(res => {
+                if (res && res.success) {
```
` if (res && res.success) {` уже не нужно проверять, так как внутри `request` все делается уже

gennady-bars commented on src/shared/services/actions/order-details.js:
```diff
@@ -11,17 +11,21 @@ export function getOrderDetails(data) {
+                        order: res
+                    });
+                } else {
+                    dispatch({
+                        type: GET_ORDER_DETAILS_FAILED
+                    });
```
Вот это нужно в `catch` только указывать

gennady-bars commented on src/shared/services/actions/burger-ingredients.js:
```diff
@@ -13,17 +13,21 @@ export function getBurgerIngredients() {
+                        ingredients: res.data
+                    });
+                } else {
+                    dispatch({
+                        type: GET_BURGER_INGREDIENTS_FAILED
+                    });
```
Вот это нужно в `catch` только указывать

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
